singlebeat/beat.py

```diff
@@ -424,16 +425,22 @@ async def wait_for_commands(self):
             self.pubsub_callback(msg)
 
     async def forward_stdout(self):
-        b = await self.sprocess.stdout.read_bytes(num_bytes=100, partial=True)
-        while len(b) > 0:
-            self.stdout_read_cb(b)
+        try:
             b = await self.sprocess.stdout.read_bytes(num_bytes=100, partial=True)
+            while len(b) > 0:
+                self.stdout_read_cb(b)
+                b = await self.sprocess.stdout.read_bytes(num_bytes=100, partial=True)
+        except:
+            logger.exception('error while forwarding to stdout')
 
     async def forward_stderr(self):
-        b = await self.sprocess.stderr.read_bytes(num_bytes=100, partial=True)
-        while len(b) > 0:
-            self.stderr_read_cb(b)
+        try:
             b = await self.sprocess.stderr.read_bytes(num_bytes=100, partial=True)
+            while len(b) > 0:
+                self.stderr_read_cb(b)
+                b = await self.sprocess.stderr.read_bytes(num_bytes=100, partial=True)
+        except:
```

Comment:
What can be the reason for not being able to read stderr/stdout of the child process ? And more importantly should we just ignore that or handle it as a fatal error.

What I'm thinking is, if we can't read the output of the child process for some reason, then there is def. something wrong going on. Instead of ignoring it, maybe crashing makes more sense.

Say we have 3 single-beat instances in 3 servers, if one of them has issues with reading stdout of the child process then it will crash and another single-beat on another server will spawn a child. So the system will be recovered (or would do a fail over). And the user will know that there is something going on, on the first server - because the instance will be dead.

Please let me know what you think ?

Reply:
The idea was to avoid any potential weird interaction between restarting of the child process and the logging. This is because we now spawn an async task instead of using the tornado utilities that were used before.

In the end it does not really matter that much. I do not know the interaction of the restart cli that much tbh.

Reply:
In the end it shouldn't matter much, I'll merge this and push a new version. We aren't silently ignoring the exception anyways, it's logging the error, so if there is some weirdness going on, it will still be visible.

> we now spawn an async task to forward stdout/stderr instead of using the tornado utilities that were used before.

We are still spawning tornado subprocess, maybe I've missed something ? 

```
            self.sprocess = tornado.process.Subprocess(cmd,
                        env=env,
                        stdin=subprocess.PIPE,
                        stdout=STREAM,
                        stderr=STREAM
                       )
```

https://github.com/ybrs/single-beat/blob/master/singlebeat/beat.py#L292

Do you have a pull request in the line for removing tornado completely ?

Reply:
About removing tornado completely: Not right now, sorry.

I was mainly working on getting dependencies bumped right now so we can move on with our core product.

Reply:
No worries. I was planning to work on it so just wanted to double check.

I've just uploaded 0.5.1 to pypi. 
 
Thanks again.